### ISSUE TITLE
Comparison operator on string_spans - null-terminated and otherwise.

### ISSUE THREAD

**Issue by Sillamacka** [martinmoene/gsl-lite#53]
While testing migrating my project from Microsoft's GSL to GSL-Lite, my tests started failing when doing string_span comparisons. This was because some string_spans were null-terminated, some were not.

Is it intended behaviour for string_spans which are identical other than one being null-terminated to be considered different?

I have been experimenting with adding an option for special string_span comparison operators that ignore the last element of the operands if they are equal to `'\0' - do you think this is a good approach?

My additions can be seen on [my fork.](https://github.com/Nehacoo/gsl-lite/commit/5c422104a4fdfae48de429aa2565afe9c2afcd9a)

Thank you!

**Comment by Nehacoo** [comment]
Another, simpler, approach would be in construction, i.e. ignore the null-terminator when constructing from string literal. This would cover most cases where the comparison gives unexpected results while keeping the current behaviour of the comparisons.

**Comment by martinmoene** [comment]
Thanks Magnus,

I'll have a look at the construction-case.

Perhaps time has come to add class `basic_string_span` instead of building `string_span` on on `span`?

**Comment by Nehacoo** [comment]
Nice! I too started thinking that ignoring the null terminator on construction is better.
With my additional comparison operator experiment, there are now four versions of the comparisons, which seems a bit too much - besides, a user might want to explicitly include the null terminator in the span and its comparisons.

**Comment by martinmoene** [comment]
Started to implement `basic_string_span`.

**Comment by martinmoene** [comment]
Hi Magnus (@Sillamacka),

I've added `basic_string_span` in branch [basic_string_span](https://github.com/martinmoene/gsl-lite/tree/basic_string_span). I both looked  at the proposal [p0123r2](http://wg21.link/p0123) and M-GSL.

Can you check if the changes fix the failures you had?

Thanks,
Martin

**Comment by Sillamacka** [comment]
Hello, that was fast!

It did fix the problems, but I had to fix some bugs in the code first: [here is the diff.](https://gist.github.com/Sillamacka/80e31004c7df3eeeab624ea16244f8a8)
Summary: `rbegin()` & `rend()` should return `reverse_iterator`
Signed/unsigned problems in `basic_string_span::subspan()`.
Default constructor for `basic_string_span` was marked `noexcept`, but default constructor for `span` was not.

And finally, when compiling with Clang-3.9, `basic_string_span`'s copy constructor SFINAE causes errors when instantiating `span<char>::span(span<const char> const &)` within the `is_convertible` check. I guess the overload should be disabled with SFINAE instead of duck typing in the body?

Curiously, GCC-6.2.0 handles it just fine. I wonder which is correct? (Both are compiled with the same version of libstdc++.)

As a workaround, I replaced the `is_convertible` test to check if the pointer types are convertible instead.

Another point: I think it would be useful to allow operator== on directly on string_spans and types which are convertible to string_span. [M-GSL does this](https://github.com/Microsoft/GSL/blob/a8c794602ab72c4800270ada74035424c184f6cb/gsl/string_span#L586), although it does not appear to be in the proposal.

Thank you for your work!

**Comment by Sillamacka** [comment]
It seems the strange Clang error is triggered when calling `gsl::to_string()` on any string_span.
Minimal example:
```
#include "gsl-lite.h"
#include <string>

int main() {
	gsl::cstring_span span = "Hello world";
	std::string str = gsl::to_string(span);
}
```

I can also confirm that it can be solved by constraining `span<T>::span( span<U> const & )` with SFINAE.

**Comment by martinmoene** [comment]
Thanks Magnus,

Summary: 
- [x] rbegin() & rend() should return reverse_iterator
- [x] [Note: split into two functions as with span]<br>Signed/unsigned problems in basic_string_span::subspan().
- [x] Default constructor for basic_string_span was marked noexcept, but default constructor for span was not.
- [x] [AIUI the workaround (and the way M-GSL goes about it) is 'just fine']<br>And finally, when compiling with Clang-3.9, basic_string_span's move & copy constructor SFINAE causes errors when instantiating span<char>::span(span<const char> const &) within the is_convertible check. I guess the overload should be disabled with SFINAE instead of duck typing in the body?
Curiously, GCC-6.2.0 handles it just fine. I wonder which is correct? (Both are compiled with the same version of libstdc++.)
As a workaround, I replaced the is_convertible test to check if the pointer types are convertible instead.
- [ ] [Had noticed this, will look into it]<br>Another point: I think it would be useful to allow operator== on directly on string_spans and types which are convertible to string_span. M-GSL does this, although it does not appear to be in the proposal.

**Comment by Sillamacka** [comment]
Excellent! I forgot to mention that I also changed the `std::basic_string const &`-taking constructor to work with non-default `char_traits` and `allocator` template parameters, perhaps you should add that to the list?

**Comment by martinmoene** [comment]
I'd noticed it though :)

**Comment by martinmoene** [comment]
@Sillamacka , I'll first release *gsl lite* with `basic_string_span` without its *enhanced* comparison as indicated below, then have a look at it:

- [ ] [Had noticed this, will look into it]<br>Another point: I think it would be useful to allow operator== on directly on string_spans and types which are convertible to string_span. M-GSL does this, although it does not appear to be in the proposal.

**Comment by Sillamacka** [comment]
Sounds good to me, explicitly constructing cstring_spans in comparisons isn't _that_ much extra work, so I think I can deal with it :D

**Comment by martinmoene** [comment]
It'll not take long probably...

**Comment by martinmoene** [comment]
Published [release 0.16.0](https://github.com/martinmoene/gsl-lite/releases/tag/v0.16.0) with enhanced string span comparison.

**Comment by Sillamacka** [comment]
Great! I think we can consider this issue resolved now.
